runtime/drivers/file/watcher.go

```diff
@@ -181,15 +181,15 @@ func (w *watcher) runInner() error {
 				continue
 			}
 
-			if e.Has(fsnotify.Create) {
+			if we.Type == runtimev1.FileEvent_FILE_EVENT_WRITE {
 				info, err := os.Stat(e.Name)
 				we.Dir = err == nil && info.IsDir()
 			}
 
 			w.buffer[path] = we
 
 			// Calling addDir after appending to w.buffer, to sequence events correctly
-			if we.Dir && e.Has(fsnotify.Create) {
+			if we.Dir && we.Type == runtimev1.FileEvent_FILE_EVENT_WRITE {
 				err = w.addDir(e.Name, true, false)
```

Comment:
I'm not sure if it's safe to call `addDir` multiple times for the same directory, you'll need to check the behavior in detail. At best, it will read all nested directories again (which might be expensive). At worst, it will double-track the directory, so the same events are emitted multiple times (which could snowball).

Reply:
Makes sense. Lemme try if merging events within a batch is enough here.